### PR TITLE
Issue #811 Support registering multiple instances of a single Service subtype

### PR DESCRIPTION
**Pull request by cljohnso**
The commits underlying this PR replace the `SupplementaryService` annotation with a `PluralService` annotation and generalize `ServiceLocator` support for registering multiple instances of a single `Service` subtype.  

Each `Service` supporting multiple instance registration is marked with the `PluralService` annotation.  The following existing `Service` interfaces are marked with `PluralService`:

* `Store.Provider`
* `AuthoritativeTier`
* `CachingTier`

`EhcacheManager` and `UserManagedCacheBuilder` are changed to select a `Store.Provider` implementation based on the results returned by a the new `Store.Provider.rank` method.  The ranking is based on the configured resources and service configurations.

In addition to the changes in `ServiceLocator` and the introduction of `PluralService`, this PR also includes updates to some unit tests needed to accommodate the multiple service support.

### CODE REVIEW

File: core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+   * @param resourceTypes the set of {@code ResourceType}s that must be supported by the provider
+   * @param serviceConfigs the collection of {@code ServiceConfiguration}s used to influence the
+   *                       selection
+   *
+   * @return the non-{@code null} {@code Store.Provider} implementation chosen
+   *
+   * @throws IllegalStateException if no suitable {@code Store.Provider} is available
+   */
+  public static Store.Provider selectStoreProvider(
+      final ServiceLocator serviceLocator, final Set<ResourceType> resourceTypes, final Collection<ServiceConfiguration<?>> serviceConfigs) {
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java]: Can't the type here be a `ServiceProvider`?

Comment (cljohnso) [inline core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java]: Yes, it could be -- but `ServiceLocator` is in the same module under 'spi' so I didn't see the point.  If you'd prefer it be generalized to `ServiceProvider<Service>` I can certainly do that.

Comment (chrisdennis) [inline core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java]: That would seem to make sense to me if it's the minimal necessary interface.

File: core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    final Collection<Store.Provider> storeProviders = serviceLocator.getServicesOfType(Store.Provider.class);
+    if (storeProviders.isEmpty()) {
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java]: This is a little redundant, but I guess you think the distinct message is helpful?  Maybe we should eliminate this, and include the set of available providers in the later ISE?

File: core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    }
+    int highRank = 0;
+    Store.Provider rankingProvider = null;
+    for (final Store.Provider provider : storeProviders) {
+      int rank = provider.rank(resourceTypes, serviceConfigs);
+      if (rank > highRank) {
+        highRank = rank;
+        rankingProvider = provider;
+      }
+    }
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java]: Could we do this as a Collections.max(storeProviders, RANKING_COMPARATOR) call, followed by a conditional on the return that the rank is above 0?

File: core/src/main/java/org/ehcache/core/spi/cache/Store.java [hunk 4]
```diff
@@ -446,6 +450,19 @@
...
 
+    /**
+     * Gets the internal ranking for the {@link Store} instances provided by this {@code Provider} of the store's
+     * ability to handle the specified resources.  A higher rank value indicates a more capable {@code Store}.
+     *
+     * @param resourceTypes the set of {@code ResourceType}s for the store to handle
+     * @param serviceConfigs the collection of {@code ServiceConfiguration} instances that may contribute
+     *                       to the ranking
+     *
+     * @return a non-negative rank indicating the ability of a {@code Store} created by this {@code Provider}
```
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/core/spi/cache/Store.java]: Is there any real reason to disallow negative ranks?

File: core/src/test/java/org/ehcache/core/UserManagedCacheTest.java [hunk 5]
```diff
@@ -100,7 +94,6 @@ private void assertFailingTransitionGoesToLowestStatus(InternalCache cache) thro
     } catch (StateTransitionException e) {
       assertThat(cache.getStatus(), is(Status.UNINITIALIZED));
     }
-
```
Comment (chrisdennis) [inline core/src/test/java/org/ehcache/core/UserManagedCacheTest.java]: Could we avoid this cleanup? (or do it in a separate commit)

File: impl/src/main/java/org/ehcache/impl/internal/store/DefaultStoreProvider.java [hunk 6]
```diff
@@ -46,10 +49,22 @@
 public class DefaultStoreProvider implements Store.Provider {
   private static final Logger LOG = LoggerFactory.getLogger(DefaultStoreProvider.class);
 
+  private static final Set<ResourceType> SUPPORTED_RESOURCE_TYPES =
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/DefaultStoreProvider.java]: This could be an enum set right?

File: osgi-test/build.gradle [hunk 7]
```diff
@@ -47,6 +47,7 @@ test {
   systemProperty 'ehcache.osgi.jar', project(':dist').jar.archivePath.getPath()
   systemProperty 'ehcache.osgi.jcache.version', parent.jcacheVersion
   systemProperty 'ehcache.osgi.slf4j.version', parent.slf4jVersion
+  systemProperty 'jta.osgi.jar', "$projectDir/lib/com.springsource.javax.transaction-1.1.0.jar"
```
Comment (chrisdennis) [inline osgi-test/build.gradle]: This is the ugliness that Louis is fixing right?  Could we stick a comment in here to that effect?

File: transactions/src/main/java/org/ehcache/transactions/xa/internal/XAStore.java [hunk 8]
```diff
@@ -656,6 +657,20 @@ public V apply(final K k) {
     private final Map<Store<?, ?>, SoftLockValueCombinedSerializerLifecycleHelper> createdStores = new ConcurrentWeakIdentityHashMap<Store<?, ?>, SoftLockValueCombinedSerializerLifecycleHelper>();
 
     @Override
+    public int rank(final Set<ResourceType> resourceTypes, final Collection<ServiceConfiguration<?>> serviceConfigs) {
```
Comment (chrisdennis) [inline transactions/src/main/java/org/ehcache/transactions/xa/internal/XAStore.java]: This seems wrong... XAStore should be doing a recursive call on StoreSupport.selectStoreProvider that is passed a StoreProvider that filters out this provider from the set of returned store providers.

File: core/src/main/java/org/ehcache/core/spi/cache/Store.java (hunk 4 above)
Comment (cljohnso) [inline core/src/main/java/org/ehcache/core/spi/cache/Store.java]: No real reason ... I just wanted a value of 0 to mean the provider didn't provide support.

File: core/src/test/java/org/ehcache/core/UserManagedCacheTest.java (hunk 5 above)
Comment (cljohnso) [inline core/src/test/java/org/ehcache/core/UserManagedCacheTest.java]: Hmmm ... not my code ... I just removed stuff from this class.

File: core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java (hunk 3 above)
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java]: It would be called once for each provider as it is currently... then once again for the actual max (to check the max rank against 0).

File: core/src/test/java/org/ehcache/core/UserManagedCacheTest.java (hunk 5 above)
Comment (chrisdennis) [inline core/src/test/java/org/ehcache/core/UserManagedCacheTest.java]: Sorry, I meant could we leave the imports in place since nothing else was modified in the class.

File: impl/src/main/java/org/ehcache/impl/internal/store/DefaultStoreProvider.java (hunk 6 above)
Comment (cljohnso) [inline impl/src/main/java/org/ehcache/impl/internal/store/DefaultStoreProvider.java]: If we can assume that anything supported by `DefaultStoreProvider` (or its successor) will be in `ResourceType.Core`, then an enum set will work.  I'll tweak this.

File: transactions/src/main/java/org/ehcache/transactions/xa/internal/XAStore.java (hunk 8 above)
Comment (cljohnso) [inline transactions/src/main/java/org/ehcache/transactions/xa/internal/XAStore.java]: Discussed with @chrisdennis -- unless we need to bundle it with this work, the associated updates will be another PR.

File: impl/src/main/java/org/ehcache/impl/internal/store/DefaultStoreProvider.java [hunk 9]
```diff
@@ -46,10 +49,22 @@
 public class DefaultStoreProvider implements Store.Provider {
```
Comment (cljohnso) [inline impl/src/main/java/org/ehcache/impl/internal/store/DefaultStoreProvider.java]: Discussed with @chrisdennis -- the functionality currently performed by `DefaultStoreProvider` can be subsumed into `CacheStore.Provider` along with changes to `CacheStore.Provider` to have it be specific about what resource/config combinations it supports.  Likely needs additional discussion but this work will be in another PR ... possibly combined with the `XAStore.Provider` changes.

File: core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java (hunk 2 above)
Comment (cljohnso) [inline core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java]: I suppose the likelihood of having a situation where **no** `Store.Provider` is available is exceedingly rare.  I can ditch this.

Do you want to itemize the available providers in the "No Store.Provider found ..." message?  If so, should this just be a `storeProviders.toString()` or something better formatted for readability?

File: core/src/test/java/org/ehcache/core/UserManagedCacheTest.java (hunk 5 above)
Comment (cljohnso) [inline core/src/test/java/org/ehcache/core/UserManagedCacheTest.java]: Yes, this could be reverted.  I had originally touched the file because of some test that was subsequently relocated (I'd put it in the wrong place).

File: core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java (hunk 2 above)
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java]: `Store.Providers.toString()` will be hard to digest, I believe the ideal is a list of `Store.Provider` class names.

File: transactions/src/main/java/org/ehcache/transactions/xa/internal/XAStore.java (hunk 8 above)
Comment (ljacomet) [inline transactions/src/main/java/org/ehcache/transactions/xa/internal/XAStore.java]: Agree with the change, propose to cover it in #798

File: impl/src/main/java/org/ehcache/impl/internal/store/DefaultStoreProvider.java [hunk 10]
```diff
@@ -46,10 +49,22 @@
...
+  private static final Set<ResourceType> SUPPORTED_RESOURCE_TYPES =
+      new HashSet<ResourceType>(Arrays.asList(ResourceType.Core.HEAP, ResourceType.Core.OFFHEAP, ResourceType.Core.DISK));
+
   private volatile ServiceProvider<Service> serviceProvider;
   private final ConcurrentMap<Store<?, ?>, Store.Provider> providersMap = new ConcurrentWeakIdentityHashMap<Store<?, ?>, Store.Provider>();
 
   @Override
+  public int rank(final Set<ResourceType> resourceTypes, final Collection<ServiceConfiguration<?>> serviceConfigs) {
+    if (SUPPORTED_RESOURCE_TYPES.containsAll(resourceTypes)) {
+      return resourceTypes.size();
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/store/DefaultStoreProvider.java]: This is valid only if bigger than `1`, otherwise it may get selected and later fail.

File: core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+      final ServiceLocator serviceLocator, final Set<ResourceType> resourceTypes, final Collection<ServiceConfiguration<?>> serviceConfigs) {
+    final Collection<Store.Provider> storeProviders = serviceLocator.getServicesOfType(Store.Provider.class);
+    if (storeProviders.isEmpty()) {
+      throw new IllegalStateException("No Store.Provider available");
+    }
+    int highRank = 0;
+    Store.Provider rankingProvider = null;
+    for (final Store.Provider provider : storeProviders) {
+      int rank = provider.rank(resourceTypes, serviceConfigs);
+      if (rank > highRank) {
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java]: What should be the behaviour in case multiple providers answer with the same rank?
Initial thinking indicates this is a bad idea.

File: impl/src/test/java/org/ehcache/impl/internal/store/tiering/CacheStoreTest.java [hunk 12]
```diff
@@ -680,6 +683,36 @@ public void testReleaseStoreFlushes () throws CacheAccessException {
     verify(cachingTierProvider, times(1)).releaseCachingTier(any(CachingTier.class));
   }
 
+  @Test
+  public void testRank() throws Exception {
+    CacheStore.Provider provider = new CacheStore.Provider();
+
+    Assert.assertThat(provider.rank(
+        Collections.<ResourceType>singleton(ResourceType.Core.DISK), Collections.<ServiceConfiguration<?>>emptyList()),
+        Matchers.is(0));
```
Comment (ljacomet) [inline impl/src/test/java/org/ehcache/impl/internal/store/tiering/CacheStoreTest.java]: More style than anything, but using static imports here helps readability for me.

File: impl/src/test/java/org/ehcache/impl/internal/store/tiering/CacheStoreTest.java [hunk 13]
```diff
@@ -680,6 +683,36 @@ public void testReleaseStoreFlushes () throws CacheAccessException {
     verify(cachingTierProvider, times(1)).releaseCachingTier(any(CachingTier.class));
   }
 
+  @Test
+  public void testRank() throws Exception {
+    CacheStore.Provider provider = new CacheStore.Provider();
+
+    Assert.assertThat(provider.rank(
+        Collections.<ResourceType>singleton(ResourceType.Core.DISK), Collections.<ServiceConfiguration<?>>emptyList()),
```
Comment (ljacomet) [inline impl/src/test/java/org/ehcache/impl/internal/store/tiering/CacheStoreTest.java]: Extracting the collections creation could help a lot test readability.

File: core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java (hunk 2 above)
Comment (cljohnso) [inline core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java]: K ... `Store.Provider` class names it is.

File: core/src/test/java/org/ehcache/core/UserManagedCacheTest.java (hunk 5 above)
Comment (cljohnso) [inline core/src/test/java/org/ehcache/core/UserManagedCacheTest.java]: This will be _removed_ in my next push of the PR.

File: impl/src/main/java/org/ehcache/impl/internal/store/DefaultStoreProvider.java (hunk 10 above)
Comment (cljohnso) [inline impl/src/main/java/org/ehcache/impl/internal/store/DefaultStoreProvider.java]: Addressing ...

File: core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java (hunk 11 above)
Comment (cljohnso) [inline core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java]: Will throw instead ... though, in practice, after the `DefaultStoreProvider` change to respond more accurately, the chance of this happening should be close to nil.

File: impl/src/test/java/org/ehcache/impl/internal/store/tiering/CacheStoreTest.java (hunk 12 above)
Comment (cljohnso) [inline impl/src/test/java/org/ehcache/impl/internal/store/tiering/CacheStoreTest.java]: Most existing references to `ResourceType.Core` aren't statically imported.  I've introduced a method that cleans this up considerably -- have a look at the newer stuff when it gets pushed and let me know if you still think a static import will help.

File: impl/src/test/java/org/ehcache/impl/internal/store/tiering/CacheStoreTest.java (hunk 13 above)
Comment (cljohnso) [inline impl/src/test/java/org/ehcache/impl/internal/store/tiering/CacheStoreTest.java]: Introduced another method to clean this up.

File: transactions/src/main/java/org/ehcache/transactions/xa/internal/XAStore.java [hunk 14]
```diff
@@ -771,6 +773,28 @@ public V apply(final K k) {
...
+      if (findSingletonAmongst(XAStoreConfiguration.class, serviceConfigs) == null) {
+        // An XAStore must be configured for use
+        return 0;
+      }
+
+      // TODO: Introduce proper Provider discovery & tracking and remove this *HACK*
+      final int underlyingRank;
+      if (resourceTypes.equals(EnumSet.of(ResourceType.Core.HEAP))) {
```
Comment (cljohnso) [inline transactions/src/main/java/org/ehcache/transactions/xa/internal/XAStore.java]: I used this hack instead of simply presuming the underlying provider (`DefaultStoreProvider`) supported the configuration.  This will need changing once `XSStore.Provider` is capable of discovering (and tracking) the proper provider.

(As a note, `DefaultStoreProvider` continues to support heap-only -- just doesn't "advertise" that fact.)

**Comment by cljohnso** [comment]
Rebased

File: api/src/main/java/org/ehcache/spi/service/PluralService.java [hunk 15]
```diff
@@ -22,9 +22,12 @@
 import java.lang.annotation.Target;
 
 /**
- * @author Alex Snaps
+ * Used to indicate that a {@link Service} subtype is permitted to have more than
+ * one concrete implementation registered with a {@link org.ehcache.spi.ServiceProvider}.
+ *
+ * @author Clifford W. Johnson
```
Comment (chrisdennis) [inline api/src/main/java/org/ehcache/spi/service/PluralService.java]: @ljacomet you just purged all the author tags right?  Do we care if they start coming back?

Comment (ljacomet) [inline api/src/main/java/org/ehcache/spi/service/PluralService.java]: That's a good point. @cljohnso Could you remove it? I believe this is the kind of information better served by SCM annotate and that it quickens documentation rot.

Comment (cljohnso) [inline api/src/main/java/org/ehcache/spi/service/PluralService.java]: Will do

File: core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+   * @param serviceConfigs the collection of {@code ServiceConfiguration}s used to influence the
+   *                       selection
+   *
+   * @return the non-{@code null} {@code Store.Provider} implementation chosen
+   *
+   * @throws IllegalStateException if no suitable {@code Store.Provider} is available or if
+   *        multiple {@code Store.Provider} implementations return the same top ranking
+   */
+  public static Store.Provider selectStoreProvider(
+      final ServiceProvider<Service> serviceLocator, final Set<ResourceType> resourceTypes, final Collection<ServiceConfiguration<?>> serviceConfigs) {
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java]: first parameter name should be `serviceProvider`

Comment (cljohnso) [inline core/src/main/java/org/ehcache/core/spi/cache/StoreSupport.java]: Changed
